app/forms/AssociateUcr.scala

```diff
@@ -48,29 +50,44 @@ object AssociateKind {
     Format[AssociateKind](Reads.StringReads.collect(JsonValidationError("error.unknown"))(lookup), Writes(kind => JsString(kind.formValue)))
 }
 
-case class AssociateUcr(kind: AssociateKind, ducr: Option[String], mucr: Option[String]) {
-  def ucr: String = ducr.orElse(mucr).get
-}
+case class AssociateUcr(kind: AssociateKind, ucr: String)
 
 object AssociateUcr {
   val formId: String = "AssociateDucr"
 
   implicit val format = Json.format[AssociateUcr]
 
-  val mapping =
-    Forms.mapping(
-      "kind" -> of[AssociateKind],
-      "ducr" -> optional(
-        text()
-          .verifying("associate.ducr.empty", nonEmpty)
-          .verifying("associate.ducr.error", isEmpty or validDucr)
-      ),
-      "mucr" -> optional(
-        text()
-          .verifying("associate.mucr.empty", nonEmpty)
-          .verifying("associate.mucr.error", isEmpty or validMucr)
-      )
-    )(AssociateUcr.apply)(AssociateUcr.unapply)
-
-  val form: Form[AssociateUcr] = Form(mapping)
+  val mapping: Mapping[AssociateUcr] = {
+    def bind(associateKind: AssociateKind, ducr: String, mucr: String): AssociateUcr =
+      associateKind match {
+        case Ducr => AssociateUcr(Ducr, ducr)
+        case Mucr => AssociateUcr(Mucr, ducr)
+      }
+
+    def unbind(value: AssociateUcr): (AssociateKind, String, String) =
+      value.kind match {
+        case Ducr => (value.kind, value.ucr, "")
+        case Mucr => (value.kind, "", value.ucr)
+      }
+
+    val nonEmptySelected = Constraint[(AssociateKind, String, String)]("ucr.present") {
+      case (associateKind: AssociateKind, ducr: String, mucr: String) =>
+        associateKind match {
+          case Ducr =>
+            if (ducr.isEmpty) Invalid("ducr.error.empty")
+            else {
+              if (validDucr(ducr)) Valid else Invalid("ducr.error.format")
+            }
+          case Mucr =>
+            if (mucr.isEmpty) Invalid("mucr.error.empty")
+            else {
+              if (validMucr(mucr)) Valid else Invalid("mucr.error.format")
+            }
```

Comment:
What do you think:
```
if (ducr.isEmpty) Invalid("ducr.error.empty")
else if (validDucr(ducr)) Valid
else Invalid("ducr.error.format")
```

The same for Mucr

Reply:
To me it looks bad - i will try to get something better

Reply:
Was in two lines - result of formatter

Reply:
I prefer @PatrykRudnicki